compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -671,15 +672,24 @@ public void visit(BLangWhile whileNode, AnalyzerData data) {
                 data.breakAsLastStatement || data.statementReturnsPanicsOrFails, true,
                 data.potentiallyInvalidAssignmentInLoopsInfo);
 
-        decrementLoopCount(data);
         data.failureHandled = failureHandled;
-        if (data.booleanConstCondition != symTable.trueType || data.breakStmtFound) {
+        if (whileNode.onFailClause == null && data.loopCount == 1 && !data.statementReturnsPanicsOrFails &&
```

Comment:
I think that checking `data.loopCount == 1` may not be correct. Consider the following scenario:
```ballerina
function value() returns int {
    int i  = 0;
    while i<10 {
        while true {
        
        }
        return 10;
    }
    
}
``` 
Here, we need to identify the `while true { }` construct as an infinite loop and raise an error for the `return 10;` statement as unreachable code. In the given example, we correctly identify the `return 10;` statement as unreachable, but we fail to recognize the while loop as an infinite loop. This works because we identify a fail or panic statement within the while loop, which is incorrect.